Flatten top stories data

dr3 opened the pull request:
Resolves #1578

**Overall change:** Flatten CPS top stories sections into a single group

**Code changes:**

- Creates a preprocessor rule to squash top stories
- Adds some unit tests to test dummy data
- Adds some snapshot tests to test real fixture data

Testing notes:
This isnt integrated yet and only has unit tests for now. 

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

Commit:
:rocket: Lets get some data up in here

Commit:
:rocket: Add squash logic

Commit:
:rocket: Update to be working

Commit:
:bath: Make it a bit safer

Commit:
:white_check_mark: More tests

Commit:
:bath: Remove unneeded fixture and update indexof to includes

dr3 commented:
~BLOCKED ON https://github.com/bbc/simorgh/pull/1590~

~Due to failing fixture data tests~

Commit:
Merge branch 'latest' into DataSquash

Commit:
Merge branch 'latest' into DataSquash

j-pendlebury commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+     * If items were collected, form into a new group
+     */
+    if (collectedItems.length > 0) {
+      const newGroup = {
```
could you rename this to `newTopStoriesGroup` or similar? Makes it more readable.

dr3 replied:
done :)

j-pendlebury commented on src/app/dataValidator/rules/topstories.test.js:
```diff
@@ -0,0 +1,134 @@
+import applySquashTopstories from './topstories';
+
+// data with the groups 'top-story' 'secondary-top-story' 'other-top-stories'
+import fixtureDataFormatOne from '../../../../data/prod/thai/frontpage/index.json';
```
Should we specify what the two formats are/how they differ? We might know why we have two types of data feeds, but others might not

dr3 replied:
Done :), moved comment to the bottom where the `describe` is

j-pendlebury reviewed (changes requested):
Just two small comments

Commit:
:pencil: Change variable name and add comment

Commit:
:pencil: tweak wording

dr3 commented:
Blocked on https://github.com/bbc/simorgh/pull/1597 for failing ci

amywalkerdev committed:
Merge branch 'latest' into DataSquash

amywalkerdev committed:
Merge branch 'latest' into DataSquash

amywalkerdev committed:
Merge branch 'latest' into DataSquash

pjlee11 commented on src/app/dataValidator/rules/__snapshots__/topstories.test.js.snap:
```diff
@@ -0,0 +1,14521 @@
+            "cpsType": "MAP",
+            "headlines": Object {
+              "headline": "จะเกิดอะไรขึ้นเมื่อยาปฏิชีวนะใช้ไม่ได้ผลอีกต่อไป",
+        
```
Do we really need 14521 lines of snapshot? 😨

dr3 replied:
I think so, it doesn't hurt and means if we change this logic we know it wont have negative side effects. Its more like a integration test

pjlee11 replied:
They are needed, just so longggggggg

pjlee11 commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+import deepClone from '../../helpers/json/deepClone';
```
Not convinced this is homed in the right place. `src/app/dataValidator/` is only for the dataValidation logic which is checking the Optimo payload format on the local dev env. Realistically in the future we will remove it from simorgh (_reminds me that I need to write that issue_). 

Whereas this is actual production code for flattening the payload after `getInitialData` and should live somewhere more suitable such as:
- `src/app/lib/`

pjlee11 replied:
I've since realised that there is also timestamp logic homed in the same place as this. I'd be happy for the outcome to be that we create an issue to move both of these out of the `dataValidator/` directory rather than the change being in this PR.

dr3 replied:
Done https://github.com/bbc/simorgh/issues/1616

pjlee11 commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+  'top-story',
+  'secondary-top-story',
+  'other-top-stories',
+];
```
```
"type": "responsive-top-stories", // group name for new services created as part of W2020 expansion
"type": "top-story",  // the 3 group names for services before the W2020 expansion 
"type": "secondary-top-story",
"type": "other-top-stories",
```

pjlee11 commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+     * Find and delete unwanted groups while collecting items
+     */
+    groups = groups.filter(group => {
+      if (group.type && squashKeys.includes(group.type)) {
```
```suggestion
      if (squashKeys.includes(group.type)) {
```

I don't think `group.type` is needed here because `squashKeys.includes(undefined)` returns `false`. If you were trying to guard against `group` being `undefined` then this also isn't achieving that because both sides of the `&&` are always evaluated even if the first fails. 

![image](https://user-images.githubusercontent.com/7791726/56302189-2d21ea80-6131-11e9-939b-4a4c65676954.png)

dr3 replied:
Done :)

pjlee11 commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+const squashTopStories = jsonRaw => {
+  const json = deepClone(jsonRaw); // make a copy so we don't corrupt the input
+  let groups = deepGet(['content', 'groups'], json);
+  let collectedItems = [];
```
```suggestion
  const collectedItems = [];
```

This doesn't need to be redeclared on [line 22](https://github.com/bbc/simorgh/pull/1596/files#diff-647ce4d4a42e13902b4fdc37c884a0bfR22) because we are just adding items to the array so can be a const. 

![image](https://user-images.githubusercontent.com/7791726/56302458-b76a4e80-6131-11e9-81be-7f8bf13c3ceb.png)

dr3 replied:
![image](https://user-images.githubusercontent.com/11341355/56304264-ab808b80-6135-11e9-8903-3721b3ac59f8.png)
:O

pjlee11 commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+     */
+    groups = groups.filter(group => {
+      if (group.type && squashKeys.includes(group.type)) {
+        collectedItems = collectedItems.concat(group.items);
```
```suggestion
        collectedItems.concat(group.items);
```

dr3 replied:
> The concat() method is used to merge two or more arrays. This method does not change the existing arrays, but instead returns a new array.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat

pjlee11 replied:
Fair enough, The screenshot above is confusion AF 😆

pjlee11 commented on src/app/dataValidator/rules/topstories.js:
```diff
@@ -0,0 +1,55 @@
+      groups.unshift(newTopStoriesGroup);
+
+      /*
+       * Override the original groups with the new ones
```
We make a [deep clone](https://github.com/bbc/simorgh/pull/1596/files#diff-647ce4d4a42e13902b4fdc37c884a0bfR12) not to change the original input but then overwrite the groups anyway. Considering the deepClone needs to process on the client and then we replace it anyway I'm not convinced the benefit of the cloning is worthwhile - unless I'm missing something.

dr3 replied:
If we dont deep clone we manipulate function arguments, which I dont think is good practice

pjlee11 replied:
Not convinced but would like to hear other peoples opinions. If we never compare the original input to the manipulated output I don't see why we should clone/maintain the original input

j-pendlebury replied:
I agree with Phil. If we're not really using the original data, then does it make sense to clone it?

pjlee11 reviewed (changes requested):
Some changes syntax changes and a query about where the code is homed, overall nice work! 

PS: Thanks for the comments in the unit tests made things very clear!

Commit:
:wrench: Update for phils review comment

pjlee11 reviewed (dismissed):
LGTM nice work 👍 

PS: would appreciate other reviewers thoughts on the cloning approach - [discussion here](https://github.com/bbc/simorgh/pull/1596#discussion_r276337993)

Commit:
:rocket: Add strapline logic and remove deepclone

Commit:
Merge branch 'latest' into DataSquash

dr3 commented:
@pjlee11 @jblaut I have removed the deep clone.

I have also added logic to collect a strapline name if found, due to a product requirement to preserve it.

j-pendlebury reviewed (approved):
![](https://media.giphy.com/media/l4FGCaauYn6rqQhWg/giphy.gif)

LGTM

Commit:
Merge branch 'latest' into DataSquash

pjlee11 reviewed (approved):
LGTM 👍

![](https://media.giphy.com/media/8ZhtXs9wCVGnlfHSb7/giphy.gif)

david-boydell commented:
Agreed that there is no scope for testing as this is not represented in any outward way yet.
